jetty-core/jetty-client/src/main/java/org/eclipse/jetty/client/HttpReceiver.java

```diff
@@ -731,12 +711,30 @@ private void processDemand()
             if (LOG.isDebugEnabled())
                 LOG.debug("Processing demand on {}", this);
 
-            if (currentChunk == null)
+            Content.Chunk current;
+            try (AutoLock ignored = lock.lock())
+            {
+                current = currentChunk;
+            }
+
+            if (current == null)
             {
-                currentChunk = HttpReceiver.this.read(true);
-                if (currentChunk == null)
+                current = HttpReceiver.this.read(true);
+                if (current == null)
```

Comment:
If there was a concurrent call to `fail()` here, you want to return `currentChunk` even if `current` is null.